storage/engine/db.cc

```diff
@@ -607,19 +611,61 @@ class DBMergeOperator : public rocksdb::MergeOperator {
 
 class DBLogger : public rocksdb::Logger {
  public:
-  DBLogger(DBLoggerFunc f)
-      : func_(f) {
+  DBLogger(bool enabled)
+      : enabled_(enabled) {
   }
   virtual void Logv(const char* format, va_list ap) {
-    // TODO(pmattis): forward to Go logging. Also need to benchmark
-    // calling Go exported methods from C++ to determine if this is
-    // too slow.
-    vfprintf(stderr, format, ap);
-    fprintf(stderr, "\n");
+    // TODO(pmattis): Benchmark calling Go exported methods from C++
+    // to determine if this is too slow.
+    if (!enabled_) {
+      return;
+    }
+
+    // First try with a small fixed size buffer.
+    char space[1024];
+
+    // It's possible for methods that use a va_list to invalidate the data in
+    // it upon use. The fix is to make a copy of the structure before using it
+    // and use that copy instead.
+    va_list backup_ap;
+    va_copy(backup_ap, ap);
+    int result = vsnprintf(space, sizeof(space), format, backup_ap);
+    va_end(backup_ap);
+
+    if ((result >= 0) && (result < sizeof(space))) {
+      rocksDBLog(space, result);
+      return;
+    }
+
+    // Repeatedly increase buffer size until it fits.
+    int length = sizeof(space);
+    while (true) {
+      if (result < 0) {
+        // Older behavior: just try doubling the buffer size.
```

Comment:
Ah, this is from StringAppendV (https://code.google.com/p/re2/source/browse/util/stringprintf.cc#9) which Google has released in numerous open source projects (re2, protobuf, s2-geometry, chrome, etc). I'd prefer to leave it as-is.